### PR TITLE
Only accept txs once synced

### PR DESCRIPTION
**Pull request by benaadams**
Is a lot of additional work on processing blocks and adding tx in the txpool that is completely irrelevant until synced. Skip this extra work.

Effects all types of sync (snap, full, archive etc)

## Changes

- Skip trying to add tx to txpool if not yet synced
- Don't process block events in txpool if not yet synced

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] Yes

#### If yes, did you write tests?

- [x] Yes

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs [hunk 1]
```diff
@@ -51,6 +58,24 @@ public ChainHeadInfoProvider(IChainHeadSpecProvider specProvider, IBlockTree blo
...
+            get
+            {
+                if (_hasSynced)
+                {
+                    return false;
+                }
+
+                long bestSuggestedNumber = _blockTree.FindBestSuggestedHeader()?.Number ?? 0;
+                long headNumberOrZero = _blockTree.Head?.Number ?? 0;
+                bool isSyncing = bestSuggestedNumber == 0 || bestSuggestedNumber > headNumberOrZero + 8;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs:73]: Can you make this extension method on block tree? I think I've seen this or similar logic in few other places before, can we unify?

File: src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs [hunk 2]
```diff
@@ -27,14 +29,19 @@ public ChainHeadInfoProvider(ISpecProvider specProvider, IBlockTree blockTree, I
         {
         }
 
-        public ChainHeadInfoProvider(IChainHeadSpecProvider specProvider, IBlockTree blockTree, IReadOnlyStateProvider stateProvider, ICodeInfoRepository codeInfoRepository)
+        public ChainHeadInfoProvider(IChainHeadSpecProvider specProvider, IBlockTree blockTree, IReadOnlyStateProvider stateProvider, ICodeInfoRepository codeInfoRepository, bool forceSynced = false)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs:32]: as `forceSynced` is only used in test maybe either make it internal init property or add this logic in derived test-only type?

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 3]
```diff
@@ -408,6 +410,8 @@ public void RemovePeer(PublicKey nodeId)
 
         public AcceptTxResult SubmitTx(Transaction tx, TxHandlingOptions handlingOptions)
         {
+            if (_headInfo.IsSyncing) return AcceptTxResult.AlreadyKnown;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:413]: Shouldn't we have a new result for this scenario?

**Comment by benaadams** [comment]
Accept local tx

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 3 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:413]: +1 for creating `AcceptTxResult.Syncing` or something like that

**Review by marcindsobczak — COMMENTED** [review]
It's a great change, but I would definitely not return `AcceptTxResult.AlreadyKnown` when syncing as it can be misleading

File: src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs]: Done

File: src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Blockchain/ChainHeadInfoProvider.cs]: Done

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: Added

File: src/Nethermind/Nethermind.Blockchain/BlockTreeExtensions.cs [hunk 4]
```diff
@@ -10,5 +10,14 @@ public static class BlockTreeExtensions
         public static ReadOnlyBlockTree AsReadOnly(this IBlockTree blockTree) => new(blockTree);
 
         public static BlockHeader? GetProducedBlockParent(this IBlockTree blockTree, BlockHeader? parentHeader) => parentHeader ?? blockTree.Head?.Header;
+
+        public static (bool isSyncing, long headNumberOrZero, long bestSuggested) IsSyncing(this IBlockTree blockTree, int maxDistanceForSynced = 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/BlockTreeExtensions.cs:14]: I think we can rename it for simplicity. Normally our head should be on Genesis by default, if it isn't there is something wrong. Or we can just make it nullable here.

```suggestion
        public static (bool isSyncing, long headNumber, long bestSuggested) IsSyncing(this IBlockTree blockTree, int maxDistanceForSynced = 0)
```

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**